tests/test_functions.py

```diff
@@ -139,6 +139,95 @@ def example():
             assert example() == 17
             """)
 
+    def test_different_globals_may_have_different_builtins(self):
+        self.assert_ok("""\
+            def replace_globals(f, new_globals):
+                import sys
+                if sys.version_info.major == 2:
+                    args = [
+                        f.func_code,
+                        new_globals,
+                        f.func_name,
+                        f.func_defaults,
+                        f.func_closure,
+                    ]
+                else:
+                    args = [
+                        f.__code__,
+                        new_globals,
+                        f.__name__,
+                        f.__defaults__,
+                        f.__closure__,
+                    ]
+                if hasattr(f, '_vm'):
+                    name = args.remove(args[2])
+                    args.insert(0, name)
+                    args.append(f._vm)
+                return type(lambda: None)(*args)
+
+
+            def f():
+                assert g() == 2
+                assert a == 1
+
+
+            def g():
+                return a  # a is in the builtins and set to 2
+
+
+            # g and f have different builtins that both provide ``a``.
+            g = replace_globals(g, {'__builtins__': {'a': 2}})
+            f = replace_globals(f, {'__builtins__': {'a': 1}, 'g': g})
+
+
+            f()
+            """)
+
+    def test_no_builtins(self):
+        self.assert_ok("""\
+            def replace_globals(f, new_globals):
+                import sys
+
+
+                if sys.version_info.major == 2:
+                    args = [
+                        f.func_code,
+                        new_globals,
+                        f.func_name,
+                        f.func_defaults,
+                        f.func_closure,
+                    ]
+                else:
+                    args = [
+                        f.__code__,
+                        new_globals,
+                        f.__name__,
+                        f.__defaults__,
+                        f.__closure__,
+                    ]
+                if hasattr(f, '_vm'):
+                    name = args.remove(args[2])
+                    args.insert(0, name)
+                    args.append(f._vm)
+                return type(lambda: None)(*args)
+
+
+            def f(NameError=NameError, AssertionError=AssertionError):
+                # capture NameError and AssertionError early because
+                #  we are deleting the builtins
+                None
```

Comment:
What does this `None` do?

Reply:
That's `None` of your business.

Reply:
I should add a comment here. The default builtins, when not inherited nor provided in the globals, is just the mapping `{'None': None}`. Having this name lookup here was supposed to test that the name `None` resolves to `None`. I believe this test is insufficient in Python 3 where `None` will compile into a `LOAD_CONST` instead of a `LOAD_GLOBAL` or `LOAD_NAME`. I am not totally sure how to test that `None` is actually in the builtins given that by definition of this test, there is no `__builtins__` object in scope.